R/data.table.R

```diff
@@ -1620,8 +1620,9 @@ replace_dot_alias = function(e) {
         (jsub[[1L]] == "[" ||
            (jsub[[1L]] == "[[" && is.name(jsub[[2L]]) && eval(call('is.atomic', jsub[[2L]]), x, parent.frame()))) &&
         (is.numeric(jsub[[3L]]) || jsub[[3L]] == ".N")
-      headopt = jsub[[1L]] == "head" || jsub[[1L]] == "tail"
-      firstopt = jsub[[1L]] == "first" || jsub[[1L]] == "last" # fix for #2030
+      headopt =   jsub[[1L]]=="head"  || jsub[[1L]]=="tail"
```

Comment:
nit: `jsub[[1L]] %chin% c("head", "tail")` works, right? ditto the next line

Reply:
Seems to be a choice between readability/clarity and speed.

```r
# first hit
x = "a"
bench::mark(
    x == "a" || x == "b",
    x %chin% c("a", "b")
)
#> # A tibble: 2 × 6
#>   expression                min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>           <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 x == "a" || x == "b"    171ns    189ns  4887689.        0B        0
#> 2 x %chin% c("a", "b")    675ns    727ns  1264336.    2.22KB        0

# second hit
x = "b"
bench::mark(
    x == "a" || x == "b",
    x %chin% c("a", "b")
)
#> # A tibble: 2 × 6
#>   expression                min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>           <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 x == "a" || x == "b"    321ns    338ns  2510554.        0B        0
#> 2 x %chin% c("a", "b")    706ns    972ns   863435.        0B        0

# no hit
x = "c"
bench::mark(
    x == "a" || x == "b",
    x %chin% c("a", "b")
)
#> # A tibble: 2 × 6
#>   expression                min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>           <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 x == "a" || x == "b"    319ns    339ns  2381597.        0B        0
#> 2 x %chin% c("a", "b")    684ns    762ns  1180392.        0B        0

Reply:
thanks! ok let's leave a TODO to try and shrink that gap.

Reply:
Filed: https://github.com/Rdatatable/data.table/issues/5334